core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -2662,6 +2662,37 @@ public static void jsonArrayAggAdd(List list, Object element,
     }
   }
 
+  public static String jsonType(Object input) {
+    String result = "";
+    try {
+      Object o = input == null ? null : dejsonize(input.toString());
+      if (o instanceof Integer) {
+        result = "INTEGER";
+      } else if (o instanceof String) {
+        result = "STRING";
+      } else if (o instanceof Float) {
+        result = "FLOAT";
+      } else if (o instanceof Double) {
+        result = "DOUBLE";
+      } else if (o instanceof Long) {
+        result = "LONG";
+      } else if (o instanceof Boolean) {
+        result = "BOOLEAN";
+      } else if (o instanceof Date) {
+        result = "DATE";
+      } else if (o instanceof Map) {
+        result = "OBJECT";
+      } else if (o instanceof Collection) {
+        result = "ARRAY";
+      } else if (o == null) {
+        result = "NULL";
+      }
+    } catch (Exception ex) {
+      throw new CalciteException(ex.getMessage(), ex.getCause());
```

Comment:
The exception should be thrown using the class RESOURCE.java, you can find some usage examples just in this file.